src/openloco/ui/dropdown.cpp

```diff
@@ -263,43 +725,162 @@ namespace openloco::ui::dropdown
+                origin.x += width + 3;
+                origin.y = 0;
+            }
+        }
+
+        size.width = width + 3;
+        common::widgets[0].right = size.width;
+        size.width++;
+
+        if (origin.x < 0)
+        {
+            origin.x = 0;
+        }
+
+        origin.x += size.width;
+
+        if (origin.x > ui::width())
+        {
+            origin.x = ui::width();
+        }
+
+        origin.x -= size.width;
 
-        call(0x4CCC7C, regs);
+        common::open(origin, size, colour);
+    }
+
+    void showText2(int16_t x, int16_t y, int16_t width, int16_t height, colour_t colour, size_t count, uint8_t flags)
+    {
+        showText2(x, y, width, height, 0, colour, count, flags & ~(1 << 6));
     }
 
     // 0x004CF2B3
     void populateCompanySelect(window* window, widget_t* widget)
     {
-        registers regs;
-        // regs.edx = widgetIndex;
-        regs.edi = (int32_t)widget;
-        regs.esi = (int32_t)window;
+        char* buffer = _byte_112CC04;
+
+        while (&buffer[0] < &_byte_112CC04[15])
+        {
+            *buffer = '\0';
+            buffer++;
+        }
+
+        int16_t maxPerformanceIndex;
+        company_id_t companyId = company_id::null;
+        auto count = 0;
+
+        while (true)
+        {
+            maxPerformanceIndex = -1;
+            for (const auto& company : companymgr::companies())
+            {
+                if (company.empty())
+                    continue;
+
+                if (_byte_112CC04[company.id()] & 1)
+                    continue;
+
+                if (maxPerformanceIndex < company.performance_index)
+                {
+                    maxPerformanceIndex = company.performance_index;
+                    companyId = company.id();
+                }
+            }
+
+            if (maxPerformanceIndex == -1)
+                break;
 
-        call(0x004CF2B3, regs);
+            _byte_112CC04[companyId] = _byte_112CC04[companyId] | 1;
+            _dropdownItemFormats[count] = string_ids::dropdown_company_select;
+            _menuOptions[count] = companyId;
+
+            auto company = companymgr::get(companyId);
+            auto competitorObj = objectmgr::get<competitor_object>(company->competitor_id);
+            auto ownerEmotion = company->owner_emotion;
+            auto imageId = competitorObj->images[ownerEmotion];
+            imageId = gfx::recolour(imageId, company->mainColours.primary);
+
+            add(count, string_ids::dropdown_company_select, { imageId, company->name });
+            count++;
+        }
+        auto x = widget->left + window->x;
+        auto y = widget->top + window->y;
+        auto colour = colour::translucent(window->colours[widget->colour]);
+
+        showText(x, y, widget->width(), widget->height(), 25, colour, count, (1 << 6));
+
+        auto highlightedIndex = company_id::null;
+        highlightedIndex++;
+
+        while (window->owner != _menuOptions[highlightedIndex])
```

Comment:
If higlightedIndex gets above 15 you will have issues this should exit and set -1 as the index if that happens.